src/ord/envelope.rs

```diff
@@ -0,0 +1,275 @@
+use std::{collections::BTreeMap, convert::TryInto};
+
+use bitcoin::{util::taproot::TAPROOT_ANNEX_PREFIX, Script, Transaction};
+#[cfg(feature="schemars")]
+use schemars::JsonSchema;
+#[cfg(feature="serde")]
+use serde_derive::{Deserialize, Serialize};
+
+use {
+    super::*,
+    bitcoin::blockdata::{
+        opcodes,
+        script::{
+            self,
+            Instruction::{self, Op, PushBytes},
+            Instructions,
+        },
+    },
+    std::iter::Peekable,
+};
+
+
+type Result<T> = std::result::Result<T, script::Error>;
+type RawEnvelope = Envelope<Vec<Vec<u8>>>;
+pub(crate) type ParsedEnvelope = Envelope<Inscription>;
+
+#[derive(Default, PartialEq, Clone, Debug, Eq)]
+#[cfg_attr(feature="serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature="schemars", derive(JsonSchema))]
+pub struct Envelope<T> {
+    pub input: u32,
+    pub offset: u32,
+    pub payload: T,
+    pub pushnum: bool,
+    pub stutter: bool,
+}
+
+impl From<RawEnvelope> for ParsedEnvelope {
+    fn from(envelope: RawEnvelope) -> Self {
+        let body = envelope
+            .payload
+            .iter()
+            .enumerate()
+            .position(|(i, push)| i % 2 == 0 && push.is_empty());
+
+        let mut fields: BTreeMap<&[u8], Vec<&[u8]>> = BTreeMap::new();
+
+        let mut incomplete_field = false;
+
+        for item in envelope.payload[..body.unwrap_or(envelope.payload.len())].chunks(2) {
+            match item {
+                [key, value] => fields.entry(key).or_default().push(value),
+                _ => incomplete_field = true,
+            }
+        }
+
+        let duplicate_field = fields.iter().any(|(_key, values)| values.len() > 1);
+
+        let content_encoding = Tag::ContentEncoding.remove_field(&mut fields);
+        let content_type = Tag::ContentType.remove_field(&mut fields);
+        let delegate = Tag::Delegate.remove_field(&mut fields);
+        let metadata = Tag::Metadata.remove_field(&mut fields);
+        let metaprotocol = Tag::Metaprotocol.remove_field(&mut fields);
+        let parent = Tag::Parent.remove_field(&mut fields);
+        let pointer = Tag::Pointer.remove_field(&mut fields);
+
```

Comment:
there's also an `unbound` tag (66). are you just having it get rolled into `unrecognized_even_field`?

Reply:
I coppied this code from the ord repo so i'd do whatever they say! is 66 good?